### DISCUSSION TITLE
WordPress and Composer

### DISCUSSION THREAD

**Discussion by jeffreyvr** [spatie/wordpress-ray#38] [Ideas]
This plugin uses Imposter to prefix all composer loaded classes. However, Imposter does not support Traits. It also seems to cause issues when using Tinkerwell.

Some issues on this repo are related to that.

I've tried alternatives:
- [PHP Scoper](https://github.com/humbug/php-scoper)
- [Mozart](https://github.com/coenjacobs/mozart)
- [Jetpack Autoloader](https://github.com/TypistTech/imposter-plugin)

They all seem to have some issues. Jetpack autoloader is the easiest to implement. Downside is that it tries to load the most recent version of class. That might cause conflicts when a package specifically needs an older version.

However, Jetpack autoloader is used within the Jetpack plugin which is active on millions of WordPress sites.

Then Mozart doesn't seem to be able to load helper functions. Check out this issue for more info: https://github.com/spatie/wordpress-ray/pull/33

Today I tried (tried it in the past, but a new attempt :)) getting Ray going with PHP Scoper. I seem to have a working example:
https://github.com/jeffreyvr/wordpress-ray/tree/scoper

Downside here: it creates a separate build folder in which the plugin is placed with the prefixes applied. A made a `build.sh` script to make executing the build proces a bit easier and apply some minor tweaks. 

Another downside is that it tries to prefix all WordPress related classes and functions as well. So if in the future, if we would want to add a class like `WP_Query' in the code of Ray for example, we would have to whitelist that class to prevent it from being prefixed.

Long story short: still not sure what the best way to go about this is. But still wanted to leave this here in case someone else wants to give it a shot.

**Comment by jeffreyvr** [comment]
Another attempt at PHP Scoper: https://github.com/jeffreyvr/wordpress-ray/tree/scoper2

This version runs PHP Scoper, but only partly uses the resulting build folder (which is then renamed to dependencies).

https://github.com/jeffreyvr/wordpress-ray/blob/scoper2/build.sh

This approach seems to fix the earlier mentioned downside (having to whitelist WordPress core functions).

I've tested the following functions:
```php
ray()->showMails();
ray()->showQueries();
ray('test');
ray(new DateTime);
```

They all work fine. The issue with Tinkerwell is resolved as well.

**Reply by freekmurze** [reply]
lollll https://github.com/jeffreyvr/wordpress-ray/blob/scoper2/build.sh#L41-L46

Feel free to PR the necessary changes to this repo. I'll test it out, and if all works correctly I'll merge and publish a new version. Make sure to update the instructions in the readme on how to create a release.
